content/post/mnist-example.md

```diff
@@ -0,0 +1,155 @@
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    ab = wallaroo.ApplicationBuilder("MNIST classification")
+    ab.new_pipeline("MNIST",
+                    wallaroo.TCPSourceConfig(in_host, in_port, decode))
+    ab.to(pca_transform)
+    ab.to(logistic_classify)
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, encode))
+    return ab.build()
+```
+
+Note how we are loading our model in `application_setup` and making them global.  This is because loading a pickled scikit-learn model could be a potentially expensive operation, and we want to ensure we do it only once per worker. By doing so in the application_setup, the models are only loaded during initialization, and made available to the computations via the `global` keyword.
+
+### Decoder
+
+We will be sending pickled images from a specialised sender application, and we will have to unpickle them in the decoder as they are received
+
+```python
+@wallaroo.decoder(header_length=4, length_fmt=">I")
+def decode(bs):
+    return pickle.loads(bs)
+```
+
+### PCA Computation
+
+The first computation applies the PCA that has been previously loaded.
+
+```python
+@wallaroo.computation(name="PCA")
+def pca_transform(x):
+    return pca.transform([x])
+```
+
+### Logistic Regression computation
+
+The second computation applies the logistic regression, also previously loaded.
+
+```python
+@wallaroo.computation(name="Logistic Regression")
+def logistic_classify(x):
+    return logistic.predict(x)
+```
+
+### Encoder
+
+The encoder packs the result into a string and sends it across the wire with a framed header containing the length of the string.
+
+```python
+@wallaroo.encoder
+def encode(data):
+    s = str(data)
+    return struct.pack('>I{}s'.format(len(s)), len(s), s)
+```
+
+## Sending data to Wallaroo
+
+In order to send data into Wallaroo, we must use a special sender that knows how to pickle the images in the right way. We can create this sender by formatting our messages such that they match the working of the decoder:
+
+* 4 bytes representing the length of the message, followed by
+* a UTF-8 encoded string, containing the pickled object
+
+Our sender will send the entire MNIST dataset.
+
+```python
+import sys
+import socket
+import struct
+from sklearn import linear_model, decomposition, datasets
+
+import cPickle as pickle
+
+def send_message(conn, x):
+    msg = pickle.dumps(x)
+    conn.sendall(struct.pack('>I', len(msg)))
+    conn.sendall(msg)
+
+if __name__ == '__main__':
+    add = sys.argv[1].split(':')
+    wallaroo_input_address = (add[0], int(add[1]))
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    print 'connecting to Wallaroo on %s:%s' % wallaroo_input_address
+    sock.connect(wallaroo_input_address)
+    digits = datasets.load_digits()
+    for x in digits.data:
+        send_message(sock, x)
+```
+
+## Running our application
+
+To run our application, we need to follow these steps:
```

Comment:
Can we elaborate on what folks will see when they run the application so they know what to expect?